man/man1/xtfs_mrcdbtool.1
.TH xtfs_mrcdbtool 1 "July 2011" "The XtreemFS Distributed File System" "XtreemFS Admin Tools"
.SH NAME
xtfs_mrcdbtool \- dumps and restores MRC databases.
.SH SYNOPSIS
\fBxtfs_mrcdbtool [\fIoptions\fB] \fIdump|restore\fP <dump_file>
.br

.SH DESCRIPTION
.I xtfs_mrcdbtool
dumps an MRC database to a file, or restores an MRC database from a dump file. An XML dump of the MRC database is created if the \fIdump\fP option is chosen. When dumping an MRC database, the XML file containing the dump will be created on the server at the path \fIdump_file\fP. Dumps can be restored by using the \fIrestore\fP option. For safety reasons, this is only possible if the target MRC does not have a database yet.

.SH EXAMPLE USAGE
.TP
.B "xtfs_mrcdbtool -mrc pbrpc://localhost:32636 dump /tmp/dump.xml"
Dumps the database of the MRC running on \fIlocalhost:32636\fP to \fI/tmp/dump.xml\fP.
.TP
.B "xtfs_mrcdbtool -mrc pbrpc://localhost:32636 restore /tmp/dump.xml"
Restores the database of the MRC running on \fIlocalhost:32636\fP from the dump in \fI/tmp/dump.xml\fP.

.SH OPTIONS
.TP
.B \-c \fIcreds_file
Path to a PKCS#12 credentials file (private key + certificate) to use for SSL authentication. Must be present when MRC URL starts with pbrpcs:// or pbrpcg://.
.TP
.B \-cpass \fIcreds_passphrase
An optional passphrase to access the credentials file.
.TP
.B \-h/--help
Shows usage info.
.TP
.B \-mrc \fImrc_url
The URL the MRC, e.g. pbrpc://localhost:32636.
.TP
.B \-t \fItrusted_CAs
Path to a PKCS#12 file containing a set of certificates from trusted certification authorities. These certificates will be used to authenticate the MRC. Must be present when MRC URL starts with pbrpcs:// or pbrpcg://.
.TP
.B \-tpass \fItrusted_passphrase
An optional passphrase to access the truststore file.
.TP
.B \--admin_password \fI<passphrase>
The administrator password, if password protection is enabled.
.TP
\fB---ssl-protocol \fI<ssl-protocol> 
SSL/TLS version to use: ssltls, tlsv1, tlsv11, tlsv12. 'ssltls' (default) accepts all versions, the others accept only the exact version they name. 'tlsv12' is available in JDK 7+ only. 'tlsv11' comes with JDK 6 or 7, depending on the vendor.

.SH "SEE ALSO"
.BR xtfs_chstatus (1),
.BR xtfs_cleanup (1),
.BR xtfs_remove_osd (1),
.BR xtfs_scrub (1),
.BR

.SH AVAILABILITY
The xtfs_mrcdbtool command is part of the XtreemFS-tools package and is available from \fIhttp://www.xtreemfs.org\fP.

CHANGELOG
26-OCT-2016: release 1.6.0 (Phantastic Pancakes)

  Changes:
    * Extended volume quota implementation to OSDs. This requires dumping and reimporting the MRC database during upgrade (see 'MRC Database Backups and Conversion' in the user guide).
      * Implemented hard quotas for volumes and user and groups per volume.
      * Added quota information to import and export function.
      * These changes supersede the previous implementation.
    * Fix for mount.xtreemfs to fully support mount options via /etc/fstab.
    * Introduced Java wrappers for the C++ version of libxtreemfs and made it the default Java implementation.
    * Improved Hadoop adapter to work with a broader range of systems and versions.
    * Added 'listen.port.bind_retries' server option that implements exponential backoff during service startup if the port is not free yet.
    * Added RabbitMQ OSD tracing policy (ID 6003).
    * Added Preferred UUID OSD selection policy (ID 1003).
    * Extended X509 authentication to allow user and group mapping.
    * Slimmed down and updated Mac OS X installer with stricter dependency checking and SIP awareness.
    * Converted the Java parts to Maven and made all components available as artifacts hosted on GitHub.
    * Cleanup of code base to exclude generated sources and previously shipped binaries.
    * Add non-interactive console support to command line tools.
    * Remove explicit support for SSLv3.
    * Many small stability improvements and fixes.

12-MAR-2015: release 1.5.1

  Changes:
    * Hadoop Adapter supports Hadoop-2.x and other applications running on the YARN platform (tested with Apache Spark and Apache Flink), new OSD selection policy that imitates the HDFS placement behavior.
    * Consistent adding and removing replicas for R/W replication.
    * xtfs_scrup now replaces failed replicas using the R/W replication policy.
    * Improvements of the SSL mode: the used SSL/TLS version is selectable, strict certificate chain checks are possible, the SSL code on client and server side was improved.
    * Support for PKCS12 certificate containers is remove from the Windows client (did not work in previous release).
    * mount.xtreemfs accepts all parameters to be passed in the form -o option=value. If XtreemFS is mounted via /etc/fstab, mount options are passed to the XtreemFS client.
    * Dockerfiles for the XtreemFS services are available at https://github.com/xtreemfs/xtreemfs-docker.
    * A Vagrantfile is available in the XtreemFS repository to ease development and testing for new contributors.
    * Fixed snapshots in the case of deleted files.
    * Snapshots can be deleted even after inactivating snapshots for a volume (issue 290).
    * Initial version of an LD_PRELOAD based client that bypasses FUSE (experimental).
    * Initial implementation of volume quotas (checked only while opening files by the MRC).
    * OSDs can report their health to the DIR (e.g. based on SMART values), which aggregates this information in the web interface. A new OSD selection policy sorts out unhealthy OSDs.
    * Improved automatic testing: introduced Travis-CI, migrated all JUnit tests to version 4, improved reliability of nightly integration tests.
    * Added new benchmark tool xtfs_benchmark that is based on the JAVA libxtreemfs.
    * Many small improvements and fixes.

12-MAR-2014: release 1.5 (Wonderful Waffles)

  Changes:
    * Improved Hadoop Adapter e.g., added support for multiple volumes and read and write buffer to speed up performance.
    * Support for SSDs: Support for multiple OSD storage threads to increase parallelism.
    * Status webpage as part of the DIR webinterface for replicated files to show current primary and backup replicas.
    * Multi-homing support. XtreemFS can now be made available for multiple networks and clients will pick the correct address automatically.
    * Support for multiple OSDs per machine (e.g. one for each disk) through the xtreemfs-osd-farm init.d script.
    * Fixed major issues in Read-Only and Read/Write file replication.
    * Reduced metadata cache default timeout from 120 to 10 seconds. Disable it completely with --metadata-cache-size=0.
    * Upgraded to Google Protobuffers 2.5 which improves performance.
    * Windows Client: Fixed several issues (e.g., 286, 299) and updated the CbFS driver to the latest version.
    * XtreemFS also compiles under buildroot now.
    * Fixed several issues around SSL (e.g., 263, 281).
    * Fixed many small issues e.g., reported volume capacity was inaccurate (issue 296).
      Other fixed issues: 267, 268, 272, 274, 276, 277, 278, 283, 287, 288, 292, 293, 294.
    * Improved general stability (e.g., see issues 297, 301).
    * xtfs_scrub now repairs replicas with an invalid checksum automatically.
    * Disabled deferred deletion of file data on the OSDs.
    * Improved test coverage of nightly tests and added missing unit tests and a Valgrind leak check for C++ unit tests.

12-NOV-2012: release 1.4 (Salty Sticks)

  Changes:
    * Improved stability: Fixed client crashes when timeouts occurred; fixed server crashes and issues with the R/W replication.
    * Full support for asynchronous writes: If a client is started with "--enable-async-writes", write() requests will be immediately acknowledged by the client and executed in the background. Outstanding writes are always flushed at close() or fsync*() events. This improves the write throughput, especially in case of connections with high latency.
    * Added Windows Client Beta which uses Eldos' Callback File System product instead of Dokan.
    * Re-wrote HDFS interface implementation: Use XtreemFS as replacement for HDFS in your Hadoop setup.
    * libxtreemfs for Java: Access XtreemFS directly from your Java application. See source files org.xtreemfs.common.libxtreemfs.{Client,Volume,FileHandle}.java for the interface.
    * Re-added Vivaldi support and added/improved its functionality:

java/servers/src/org/xtreemfs/utils/xtfs_mrcdbtool.java
/*
 * Copyright (c) 2009-2011 by Bjoern Kolbeck, Jan Stender,
 *                            Zuse Institute Berlin
 *
 * Licensed under the BSD License, see LICENSE file for details.
 *
 */

package org.xtreemfs.utils;

import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.xtreemfs.foundation.SSLOptions;
import org.xtreemfs.foundation.TimeSync;
import org.xtreemfs.foundation.logging.Logging;
import org.xtreemfs.foundation.pbrpc.Schemes;
import org.xtreemfs.foundation.pbrpc.client.PBRPCException;
import org.xtreemfs.foundation.pbrpc.client.RPCAuthentication;
import org.xtreemfs.foundation.pbrpc.client.RPCNIOSocketClient;
import org.xtreemfs.foundation.pbrpc.client.RPCResponse;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.Auth;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.AuthPassword;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.AuthType;
import org.xtreemfs.foundation.pbrpc.generatedinterfaces.RPC.POSIXErrno;
import org.xtreemfs.foundation.util.CLIParser;
import org.xtreemfs.foundation.util.CLIParser.CliOption;
import org.xtreemfs.pbrpc.generatedinterfaces.GlobalTypes.PORTS;
import org.xtreemfs.pbrpc.generatedinterfaces.MRCServiceClient;

public class xtfs_mrcdbtool {
    
    /**
     * @param args
     */
    public static void main(String[] args) {
        
        Logging.start(Logging.LEVEL_ERROR);
        try {
            TimeSync.initializeLocal(0);
        } catch (Exception ex) {
            ex.printStackTrace();
            System.exit(1);
        }
        
        Map<String, CliOption> options = utils.getDefaultAdminToolOptions(true);
        List<String> arguments = new ArrayList<String>(3);
        CliOption oMrc = new CliOption(CliOption.OPTIONTYPE.URL,
            "MRC to use (e.g. 'pbrpc://localhost:32636')", "<uri>");
        oMrc.urlDefaultPort = PORTS.MRC_PBRPC_PORT_DEFAULT.getNumber();
        oMrc.urlDefaultProtocol = Schemes.SCHEME_PBRPC;
        options.put("mrc", oMrc);
        
        try {
            CLIParser.parseCLI(args, options, arguments);
        } catch (Exception exc) {
            System.out.println(exc);
            usage(options);
            System.exit(1);
        }
        
        CliOption h = options.get(utils.OPTION_HELP);
        CliOption help = options.get(utils.OPTION_HELP_LONG);
        if (h.switchValue || help.switchValue) {
            usage(options);
            System.exit(0);
        }
        
        CliOption mrc = options.get("mrc");
        if (mrc.urlValue == null) {
            System.out.println("missing MRC URL");
            usage(options);
            System.exit(1);
        }
        
        if (arguments.size() < 2) {
            usage(options);
            System.exit(1);
        }
        
        String op = arguments.get(0);
        if (!"dump".equals(op) && !"restore".equals(op)) {
            System.out.println("invalid operation: " + op);
            usage(options);
            System.exit(1);
        }
        
        String dumpFile = arguments.get(1);
        
        CliOption c = options.get(utils.OPTION_USER_CREDS_FILE);
        String cp = options.get(utils.OPTION_USER_CREDS_PASS).stringValue;
        if(cp != null && cp.equals("-")) {
        	cp = new String(System.console().readPassword("Enter credentials password: "));
        }
        CliOption t = options.get(utils.OPTION_TRUSTSTORE_FILE);
        String tp = options.get(utils.OPTION_TRUSTSTORE_PASS).stringValue;
        if(tp != null && tp.equals("-")) {
        	tp = new String(System.console().readPassword("Enter trust store password: "));
        }
        
        String sslProtocolString = options.get(utils.OPTION_SSL_PROTOCOL).stringValue;
        
        String host = mrc.urlValue.getHost();
        int port = mrc.urlValue.getPort();
        String protocol = mrc.urlValue.getProtocol();
        
        RPCNIOSocketClient rpcClient = null;
        
        try {
            SSLOptions sslOptions = null;
            boolean gridSSL = false;
            if (protocol.startsWith(Schemes.SCHEME_PBRPCS) || protocol.startsWith(Schemes.SCHEME_PBRPCG)) {
                if (c.stringValue == null) {
                    System.out.println("SSL requires '-" + utils.OPTION_USER_CREDS_FILE + "' parameter to be specified");
                    usage(options);
                    System.exit(1);
                } else if (t.stringValue == null) {
                    System.out.println("SSL requires '-" + utils.OPTION_TRUSTSTORE_FILE + "' parameter to be specified");
                    usage(options);
                    System.exit(1);
                }
                
                if (protocol.startsWith(Schemes.SCHEME_PBRPCG)) {
                    gridSSL = true;
                }

                sslOptions = new SSLOptions(c.stringValue, cp, SSLOptions.PKCS12_CONTAINER, t.stringValue, tp,
                        SSLOptions.JKS_CONTAINER, false, gridSSL, sslProtocolString, null);
            }
            rpcClient = new RPCNIOSocketClient(sslOptions, Integer.MAX_VALUE - 1000, Integer.MAX_VALUE, "xtfs_mrcdbtool");
            rpcClient.start();
            MRCServiceClient client = new MRCServiceClient(rpcClient, new InetSocketAddress(host, port));
            
            Auth passwdAuth = RPCAuthentication.authNone;
            if (options.get(utils.OPTION_ADMIN_PASS).stringValue != null)
                passwdAuth = Auth.newBuilder().setAuthType(AuthType.AUTH_PASSWORD).setAuthPasswd(
                    AuthPassword.newBuilder().setPassword(options.get(utils.OPTION_ADMIN_PASS).stringValue)).build();
            
            if (op.equals("dump")) {
                RPCResponse<?> r = null;
                try {
                    r = client.xtreemfs_dump_database(null, passwdAuth, RPCAuthentication.userService,
                        dumpFile);
                    r.get();
                } finally {
                    if (r != null)
                        r.freeBuffers();
                }
            } else if (op.equals("restore")) {
                RPCResponse<?> r = null;
                try {
                    r = client.xtreemfs_restore_database(null, passwdAuth, RPCAuthentication.userService,
                        dumpFile);
                    r.get();
                } finally {
                    if (r != null)
                        r.freeBuffers();
                }
            } else {
                usage(options);
                System.exit(1);
            }
            
        } catch (PBRPCException exc) {
            if (exc.getPOSIXErrno() == POSIXErrno.POSIX_ERROR_EACCES) {
                System.out.println("permission denied: admin password invalid or volumes exist already");
            } else {
                exc.printStackTrace();
            }
        } catch (Exception exc) {
            exc.printStackTrace();
        } finally {
            if (rpcClient != null) {
                rpcClient.shutdown();
            }
        }
    }
    
    public static void usage(Map<String, CliOption> options) {
        System.out.println("usage: xtfs_mrcdbtool [options] dump|restore <dump_file>\n");
        System.out.println("  " + "<dump_file> the file for the dump\n");
        System.out.println("  " + "options:");
        
        utils.printOptions(options);
    }
    
}
